pennylane/tensorbox/tensorbox.py

```diff
@@ -0,0 +1,253 @@
+# limitations under the License.
+"""This module contains the TensorBox abstract base class."""
+# pylint: disable=import-outside-toplevel
+import abc
+
+
+class TensorBox(abc.ABC):
+    """A container for array-like objects that allows array manipulation to be performed in a
+    unified manner for supported linear algebra packages.
+
+    Args:
+        tensor (array_like): instantiate the ``TensorBox`` container with an array-like object
+
+    .. warning::
+
+        The :class:`TensorBox` class is designed for internal use **only**, to ensure that
+        PennyLane templates, cost functions, and optimizers retain differentiability
+        across all supported interfaces.
+
+        For front-facing usage, consider using an established package such as
+        `EagerPy <https://github.com/jonasrauber/eagerpy>`_.
+
+
+    By wrapping array-like objects in a ``TensorBox`` class, array manipulations are
+    performed by simply chaining method calls. Under the hood, the method call is dispatched
+    to the corresponding linear algebra library based on the wrapped array type, without
+    the need to import any external libraries manually. As a result, autodifferentiation is
+    preserved where needed.
+
+    Currently, the following linear algebra packages are supported:
+
+    * NumPy
+    * Autograd
+    * TensorFlow
+    * Torch
+
+    **Example**
+
+    While this is an abstract base class, this class may be 'instantiated' directly;
+    by overloading ``__new__``, the tensor argument is inspected, and the correct subclass
+    is returned:
+
+    >>> x = tf.Variable([0.4, 0.1, 0.5])
+    >>> y = TensorBox(x)
+    >>> print(y)
+    >>> type(y)
+    <TensorBox <tf.Variable 'Variable:0' shape=(3,) dtype=float32, numpy=array([0.4, 0.1, 0.5], dtype=float32)>>
+
+    The original tensor is available via the :meth:`~.unbox()` method:
+
+    >>> y.unbox()
+    <tf.Variable 'Variable:0' shape=(3,) dtype=float32, numpy=array([0.4, 0.1, 0.5], dtype=float32)>
+
+    In addition, this class defines various abstract methods that all subclasses
+    must define. These methods allow for common manipulations and
+    linear algebra transformations without the need for importing.
+
+    >>> y.ones_like()
+    tf.Tensor([1. 1. 1.], shape=(3,), dtype=float32)
+
+    Unless specified, the returned tensors are also unified tensors, allowing
+    for method chaining:
+
+    >>> y.ones_like().expand_dims(0)
+    tf.Tensor([[1. 1. 1.]], shape=(1, 3), dtype=float32)
+    """
+
+    def __new__(cls, tensor):
+        if isinstance(tensor, TensorBox):
+            return tensor
+
+        if cls is not TensorBox:
+            return super(TensorBox, cls).__new__(cls)
+
+        namespace = tensor.__class__.__module__.split(".")[0]
+
+        if isinstance(tensor, (list, tuple)) or namespace == "numpy":
+            from .numpy_box import NumpyBox
+
+            return NumpyBox.__new__(NumpyBox, tensor)

```

Comment:
So if it is a "normal" numpy array we use autograd, but if it is a list or tuple we don't because we assume it is not diffable anyways? 

A bit tangential here but I started wondering how clear this is to new users...

Reply:
Not quite:

* If the input is a list, tuple, or `np.ndarray`, we create a `NumpyBox`, which uses vanilla numpy for all manipulations.

* If the input is a PennyLane tensor, we create an `AutogradBox`.

The reason being that autograd (a) has a bit of overhead compared to standard numpy, so no point using it when we don't have to, and (b) autograd is a bit restricted - not all numpy functions are supported, so sometimes you must do the manipulation in python rather than numpy (which adds more slow down)